### PR TITLE
prop_conv_solvert is not a messaget

### PR DESCRIPTION
**Pull request by peterschrammel**
Gets a log member instead.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/solvers/flattening/arrays.cpp [hunk 1]
```diff
@@ -48,7 +51,7 @@ literalt arrayst::record_array_equality(
   // check types
   if(op0.type() != op1.type())
   {
-    error() << equality.pretty() << messaget::eom;
+    log.error() << equality.pretty() << messaget::eom;
```
Comment (martin-cs) [inline src/solvers/flattening/arrays.cpp]: Should we move the diagnostic into the invariant?  There are many other instances of this in this file.

File: src/solvers/flattening/boolbv_constant.cpp [hunk 2]
```diff
@@ -33,8 +33,9 @@ bvt boolbvt::convert_constant(const constant_exprt &expr)
 
       if(tmp.size()!=op_width)
       {
-        error().source_location=expr.find_source_location();
-        error() << "convert_constant: unexpected operand width" << eom;
+        log.error().source_location = expr.find_source_location();
+        log.error() << "convert_constant: unexpected operand width"
+                    << messaget::eom;
         throw 0;
```
Comment (martin-cs) [inline src/solvers/flattening/boolbv_constant.cpp]: Could this become and invariant?  Likewise in the rest of the file.

File: src/solvers/flattening/boolbv_with.cpp [hunk 3]
```diff
@@ -19,15 +19,15 @@ bvt boolbvt::convert_with(const exprt &expr)
 {
   if(expr.operands().size()<3)
   {
-    error().source_location=expr.find_source_location();
-    error() << "with takes at least three operands" << eom;
+    log.error().source_location = expr.find_source_location();
+    log.error() << "with takes at least three operands" << messaget::eom;
     throw 0;
```
Comment (martin-cs) [inline src/solvers/flattening/boolbv_with.cpp]: Similar.

File: src/solvers/flattening/bv_minimize.h [hunk 4]
```diff
@@ -22,18 +22,19 @@ Purpose: Find a satisfying assignment that minimizes a given set
...
+  explicit bv_minimizet(boolbvt &_boolbv, message_handlert &message_handler)
+    : boolbv(_boolbv), log(message_handler)
   {
   }
 
   void operator()(const minimization_listt &objectives);
 
 protected:
   boolbvt &boolbv;
+  messaget log;
```
Comment (martin-cs) [inline src/solvers/flattening/bv_minimize.h:37]: Instance or reference?

File: src/solvers/strings/string_constraint.cpp [hunk 5]
```diff
@@ -22,7 +22,7 @@ static bool cannot_be_neg(const exprt &expr)
   satcheck_no_simplifiert sat_check(null_message_handler);
   symbol_tablet symbol_table;
   namespacet ns(symbol_table);
-  boolbvt solver(ns, sat_check);
+  boolbvt solver(ns, sat_check, null_message_handler);
```
Comment (martin-cs) [inline src/solvers/strings/string_constraint.cpp]: Is `null_message_handler` really the best option here?

**Review by martin-cs — APPROVED** [review]
Basically, yes but while we're changing things nearby it might be good to update some of the checks and error handling.

File: src/solvers/strings/string_constraint.cpp (hunk 5 above)
Comment (tautschnig) [reply src/solvers/strings/string_constraint.cpp]: 1) This is an internal call to check consistency of some subexpressions, so I think it's ok. (The same rationale that I had used when inserting `null_message_handler` a few lines above.)
2) Using a different message handler yields a diff of 400+ lines, because a lot of procedures in the string solver are free functions, and therefore almost each and every procedure needs to take an additional parameter.

File: src/solvers/flattening/boolbv_constant.cpp (hunk 2 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_constant.cpp]: If changing this, we should just pass a `constant_exprt` as the argument and thus use the common expression validation code via `to_constant_expr` at the call site.

File: src/solvers/flattening/boolbv_with.cpp (hunk 3 above)
Comment (tautschnig) [reply src/solvers/flattening/boolbv_with.cpp]: As above: use `to_with_expr` at the call site and get rid of this.

File: src/goto-checker/counterexample_beautification.h [hunk 6]
```diff
@@ -21,9 +21,8 @@ Author: Daniel Kroening, kroening@kroening.com
 class counterexample_beautificationt
 {
 public:
-  virtual ~counterexample_beautificationt()
-  {
-  }
+  explicit counterexample_beautificationt(message_handlert &message_handler);
+  virtual ~counterexample_beautificationt();
```
Comment (tautschnig) [inline src/goto-checker/counterexample_beautification.h]: `= default;`?

File: src/solvers/flattening/bv_minimize.cpp [hunk 7]
```diff
@@ -56,8 +56,7 @@ void bv_minimizet::operator()(const minimization_listt &symbols)
 {
   // build bit-wise objective function
 
-  prop_minimizet prop_minimize(boolbv);
-  prop_minimize.set_message_handler(get_message_handler());
+  prop_minimizet prop_minimize(boolbv, log.get_message_handler());
```
Comment (tautschnig) [inline src/solvers/flattening/bv_minimize.cpp:59]: Since you're doing this elsewhere: use `prop_minimize{...}` (applies several times in this PR).

File: src/solvers/flattening/bv_minimize.h [hunk 8]
```diff
@@ -22,18 +22,19 @@ Purpose: Find a satisfying assignment that minimizes a given set
 
 typedef std::set<exprt> minimization_listt;
 
-class bv_minimizet:public messaget
+class bv_minimizet
 {
 public:
-  explicit bv_minimizet(boolbvt &_boolbv):
-    boolbv(_boolbv)
+  explicit bv_minimizet(boolbvt &_boolbv, message_handlert &message_handler)
```
Comment (tautschnig) [inline src/solvers/flattening/bv_minimize.h]: `explicit` no longer needed.

File: unit/solvers/bdd/miniBDD/miniBDD.cpp [hunk 9]
```diff
@@ -188,9 +188,10 @@ SCENARIO("miniBDD", "[core][solver][miniBDD]")
 {
   GIVEN("A bdd for x&!x")
   {
+    null_message_handlert mh;
```
Comment (tautschnig) [inline unit/solvers/bdd/miniBDD/miniBDD.cpp]: You could as well use the common `null_message_handler` available in unit tests.

**Review by tautschnig — APPROVED** [review]

File: src/solvers/flattening/bv_minimize.h (hunk 4 above)
Comment (smowton) [reply src/solvers/flattening/bv_minimize.h:37]: No, generally in the cbmc codebase the `message_handlert` is lent by the creator but not the `messaget`. My guess is we suppose in the future the `messaget` instance might supply some attribution (e.g. class name, file name) to identify log message source.

File: src/solvers/prop/minimize.h [hunk 10]
```diff
@@ -76,10 +69,12 @@ class prop_minimizet:public messaget
   objectivest objectives;
 
 protected:
-  unsigned _iterations;
-  std::size_t _number_satisfied, _number_objectives;
-  weightt _value;
+  unsigned _iterations = 0;
```
Comment (smowton) [inline src/solvers/prop/minimize.h:72]: Suggest this unrelated cleanup belongs in a different commit

File: jbmc/src/jbmc/bmc.cpp [hunk 11]
```diff
@@ -204,7 +204,7 @@ safety_checkert::resultt bmct::stop_on_fail()
     if(options.get_bool_option("trace"))
     {
       if(options.get_bool_option("beautify"))
-        counterexample_beautificationt()(
+        counterexample_beautificationt{ui_message_handler}(
```
Comment (peterschrammel) [inline jbmc/src/jbmc/bmc.cpp:209]: `{}` is not just for beauty here - `()` is considered ambiguous and won't compile.

File: src/solvers/flattening/bv_minimize.h (hunk 4 above)
Comment (peterschrammel) [reply src/solvers/flattening/bv_minimize.h:37]: Instance

File: src/solvers/strings/string_constraint.cpp (hunk 5 above)
Comment (martin-cs) [reply src/solvers/strings/string_constraint.cpp]: Yikes!  This is a bit concerning.  Is this because of the "free function" approach or simply the way it has been used?

File: jbmc/src/jbmc/bmc.cpp (hunk 11 above)
Comment (tautschnig) [reply jbmc/src/jbmc/bmc.cpp:209]: Fair, but the coding guidelines also suggest it should be used everywhere...

File: src/solvers/strings/string_constraint.cpp (hunk 5 above)
Comment (tautschnig) [reply src/solvers/strings/string_constraint.cpp]: The free-function approach applied to a lot tightly coupled functions.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8982a402b12b4726f57e99a727148d7e987acd8e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106016604
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 61493235d468614a82a5a92552abca846e4eac46).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106028074
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
